server/src/main/java/org/elasticsearch/injection/spec/AmbiguousSpec.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.injection.spec;
+
+import java.util.stream.Stream;
+
+/**
+ * Indicates that there are multiple ways to inject a given type.
+ *
+ * <p>
+ * This is an error only if some type attempts to inject {@link #requestedType()},
+ * because there is no single best object of that type.
+ *
+ * <p>
+ * This is arranged as a tree so it can be instantiated in constant time;
+ * if there are more than two candidates, than {@link #option1} and/or {@link #option2}
+ * will themselves be an {@link AmbiguousSpec}.
+ */
+public record AmbiguousSpec(Class<?> requestedType, InjectionSpec option1, InjectionSpec option2) implements InjectionSpec {
```

Comment:
Shouldn't ambiguity be an error? But for this PR, is this actually needed?

Reply:
That's covered in the javadocs:

```
This is an error only if some type attempts to inject `requestedType()`, because there is no single best object of that type.
```

For example, we can anticipate that `LocalNodeMasterListener` will be ambiguous but that's not an error unless someone tries to inject it.

Reply:
If I remove this in this PR I'd need to add another way to report ambiguities.

Reply:
Alright alright, I did it. 🤪